url20b0f880a5_jgraph_legacy_jgraph5_tgz-pJ8-org_jgraph_example_HelloWorldJ8/src/org/jgraph/plaf/basic/BasicGraphUI.java

```diff
@@ -0,0 +1,3220 @@
+					Rectangle dirtyRegion = new Rectangle((int)(r.getX()
+							- hsize), (int)(r.getY() - hsize),
+							(int)(r.getWidth() + 2 * hsize), (int)(r.getHeight() + 2 * hsize));
+					graph.repaint(dirtyRegion);
+				}
+			} else {
+				Rectangle dirtyRegion = new Rectangle(graph.getSize());
+				graph.addOffscreenDirty(dirtyRegion);
+				graph.repaint();
+			}
+		}
+	}
+
+	/**
+	 * Listener responsible for getting cell editing events and updating the
+	 * graph accordingly.
+	 */
+	public class CellEditorHandler implements CellEditorListener, Serializable {
+		/** Messaged when editing has stopped in the graph. */
+		public void editingStopped(ChangeEvent e) {
+			completeEditing(false, false, true);
+		}
+
+		/** Messaged when editing has been canceled in the graph. */
+		public void editingCanceled(ChangeEvent e) {
+			completeEditing(false, false, false);
+		}
+	} // BasicGraphUI.CellEditorHandler
+
+	/**
+	 * This is used to get mutliple key down events to appropriately generate
+	 * events.
+	 */
+	public class KeyHandler extends KeyAdapter implements Serializable {
+		/** Key code that is being generated for. */
+		protected Action repeatKeyAction;
+
+		/** Set to true while keyPressed is active. */
+		protected boolean isKeyDown;
+
+		public void keyPressed(KeyEvent e) {
+			if (graph != null && graph.hasFocus() && graph.isEnabled()) {
+				KeyStroke keyStroke = KeyStroke.getKeyStroke(e.getKeyCode(), e
+						.getModifiers());
+
+				if (graph.getConditionForKeyStroke(keyStroke) == JComponent.WHEN_FOCUSED) {
+					ActionListener listener = graph
+							.getActionForKeyStroke(keyStroke);
+
+					if (listener instanceof Action)
+						repeatKeyAction = (Action) listener;
+					else
+						repeatKeyAction = null;
+				} else {
+					repeatKeyAction = null;
+					if (keyStroke.getKeyCode() == KeyEvent.VK_ESCAPE) {
+						// System.out.println("Calling release on " + marquee);
+						if (marquee != null)
+							marquee.mouseReleased(null);
+						if (mouseListener != null)
+							mouseListener.mouseReleased(null);
+						updateHandle();
+						graph.refresh();
+					}
+				}
+				if (isKeyDown && repeatKeyAction != null) {
+					repeatKeyAction.actionPerformed(new ActionEvent(graph,
+							ActionEvent.ACTION_PERFORMED, ""));
+					e.consume();
+				} else
+					isKeyDown = true;
+			}
+		}
+
+		public void keyReleased(KeyEvent e) {
+			isKeyDown = false;
+		}
+
+	} // End of BasicGraphUI.KeyHandler
+
+	/**
+	 * TreeMouseListener is responsible for updating the selection based on
+	 * mouse events.
+	 */
+	public class MouseHandler extends MouseAdapter implements
+			MouseMotionListener, Serializable {
+
+		/* The cell under the mousepointer. */
+		protected CellView cell;
+
+		/* The object that handles mouse operations. */
+		protected Object handler;
+
+		protected transient Cursor previousCursor = null;
+
+		/**
+		 * Invoked when a mouse button has been pressed on a component.
+		 */
+		public void mousePressed(MouseEvent e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 26. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.